modules/src/ics07_tendermint/client_def.rs

```diff
@@ -25,17 +36,117 @@ impl ClientDef for TendermintClient {
 
-        // TODO: Additional verifications should be implemented here.
+        // check that the versions of the client state and the header match
+        if client_state.latest_height.revision_number != header.height().revision_number {
+            return Err(Error::mismatched_revisions(
+                client_state.latest_height.revision_number,
+                header.height().revision_number,
+            )
+            .into());
+        };
+
+        let trusted_consensus_state = match ctx.consensus_state(&client_id, header.trusted_height) {
+            Some(ts) => downcast!(
+                ts => AnyConsensusState::Tendermint
+            )
+            .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))?,
+            None => {
+                return Err(Ics02Error::consensus_state_not_found(
+                    client_id,
+                    header.trusted_height,
+                ))
+            }
+        };
+
+        if header.height() == header.trusted_height.increment() {
+            //adjacent
+
+            // check that the header's trusted validator set is
+            // the next_validator_set of the trusted consensus state
+            if Set::hash(&header.validator_set) != trusted_consensus_state.next_validators_hash {
+                return Err(Error::invalid_validator_set(
+                    trusted_consensus_state.next_validators_hash,
+                    Set::hash(&header.validator_set),
+                )
+                .into());
+            }
+
+            // check that the validators that sign the commit of the untrusted header
+            // have 2/3 of the voting power of the current validator set.
+            if let Err(e) = voting_power_in(
+                &header.signed_header,
+                &header.validator_set,
+                TrustThresholdFraction::TWO_THIRDS,
+            ) {
+                return Err(Error::insufficient_voting_power(e.to_string()).into());
+            }
+        } else {
+            //Non-adjacent
+
+            //check that a subset of the trusted validator set, having 1/3 of the voting power
+            //signes the commit of the untrusted header
+            if let Err(e) = voting_power_in(
+                &header.signed_header,
+                &header.trusted_validator_set,
+                TrustThresholdFraction::default(),
+            ) {
+                return Err(Error::not_enough_trusted_vals_signed(e.to_string()).into());
+            };
+
+            // check that the validators that sign the commit of the untrusted header
+            // have 2/3 of the voting power of the current validator set.
+            if let Err(e) = voting_power_in(
+                &header.signed_header,
+                &header.validator_set,
+                TrustThresholdFraction::TWO_THIRDS,
+            ) {
+                return Err(Error::insufficient_voting_power(e.to_string()).into());
+            };
+        }
```

Comment:
All these checks should happen in the `tendermint-rs` light client API. It should also include the following checks:
- trusted header is not expired, i.e. it is within trusting period (note that we cannot provide a full trusted header to the API, just what we can get from the consensus state: height, time and next val hash) 
- update header timestamp is not beyond `now()` plus clock drift

Probably most of the monotonicity check between trusted and update headers should happen in the tendermint light client API. The only one left in this function should be between the update and next higher headers, just check the timestamps are monotonically increased and nothing else.